HW2/P4/driver.py

```diff
@@ -20,13 +20,95 @@ def py_median_3x3(image, iterations=10, num_threads=1):
+    #             # check if thread is 0 so that we don't check thread-1
+    #             if thread_index == 0:
+    #                 even_odd_list[(current_iteration-1)%2][1].wait()
+    #             # check if last thread so that we don't go over bound of threads list
+    #             elif thread_index == num_threads-1:
+    #                 even_odd_list[(current_iteration-1)%2][num_threads-2].wait()
+    #             else:
+    #                 # wait for thread-1 to be done with iteration-1
+    #                 even_odd_list[(current_iteration-1)%2][thread_index-1].wait()
+    #                 # wait for thread+1 to be done with iteration+1
+    #                 even_odd_list[(current_iteration-1)%2][thread_index+1].wait()
+    #         # run function with thread_index as first row and step of num_threads
+    #         filtering.median_3x3(tmpA, tmpB, thread_index, num_threads)
+    #         # swap buffers
+    #         tmpA, tmpB = tmpB, tmpA
+    #         # fire current event
+    #         even_odd_list[current_iteration%2][thread_index].set()
+    #         # if the other threads are done with this thread's previous iteration,
+    #         # clear that iteration, index event so that it can be used in the future
+    #         neighbors_done_with_previous_iteration = True
+    #         for neighbor in get_neighbors(thread_index):
+    #             if not even_odd_list[current_iteration%2][neighbor].is_set():
+    #                 neighbors_done_with_previous_iteration = False
+    #                 break
+    #         if neighbors_done_with_previous_iteration:
+    #             even_odd_list[(current_iteration-1)%2][thread_index].clear()
+
+    def threadFilter(thread_index, tmpA, tmpB):
+        for current_iteration in range(iterations):
+            # if this is the first iteration or there is only one thread,
+            # we wouldn't have to wait for anything
+            if current_iteration > 0 and num_threads >= 2:
+                # check if thread is 0 so that we don't check thread-1
+                if thread_index == 0:
+                    iteration_event_list[current_iteration-1][1].wait()
+                # check if last thread so that we don't go over bound of threads list
+                elif thread_index == num_threads-1:
+                    iteration_event_list[current_iteration-1][num_threads-2].wait()
+                else:
+                    # wait for thread-1 to be done with iteration-1
+                    iteration_event_list[current_iteration-1][thread_index-1].wait()
+                    # wait for thread+1 to be done with iteration+1
+                    iteration_event_list[current_iteration-1][thread_index+1].wait()
+            # run function with thread_index as first row and step of num_threads
+            filtering.median_3x3(tmpA, tmpB, thread_index, num_threads)
+            # swap buffers
+            tmpA, tmpB = tmpB, tmpA
```

Comment:
Maybe I'm missing something, but it seems that in each iteration all threads are swapping tmpB and tmpA independently? Don't you need to make sure the swapping is done **only once** at global level and at end of each iteration (i.e. after all threads finish for that iteration)? Doing it this way does not appear logically equivalent to the median filtering in serial version.